src/monitors/ProfileMonitor.v3

```diff
@@ -0,0 +1,184 @@
+    }
+
+    def render(out: TraceBuilder, m: Module, depth: int, total: int, maxDepth: int) {
+        var totalPercentTime = if(depth == 1, totalTime, total);
+        var nestedTime: int = totalTime;
+        var selfTime: int = totalTime;
+
+        var childArray = Lists.toArray(children);
+        childArray = Arrays.sort(childArray, 0, childArray.length, byTime);
+
+        if (maxDepth == -1 || depth < maxDepth) {
+            var childTotalTime = 0;
+            for (i < childArray.length) {
+                childTotalTime += childArray[i].totalTime;
+            }
+            selfTime = nestedTime - childTotalTime;
+        } else {
+            selfTime = -1;
+        }
+        
+        if (func == null) {
+            // Root node in ProfileTree
+            renderHeader(out);
+        } else {
+            var l = out.length;
+            var f = func.decl;
+            for (i < depth) out.puts("  ");
+            out.puts(TermColors.CYAN);
+            out.puts("func ");
+            f.render(m.names, out);
+            out.puts(TermColors.DEFAULT);
+            out.pad(' ', l + 50);
+            TraceUtil.renderRight(out, 12, TraceUtil.renderUs(_, selfTime));
+            TraceUtil.renderRight(out, 12, TraceUtil.renderUs(_, nestedTime));
+            TraceUtil.renderRight(out, 10, TraceUtil.renderPct(_, selfTime, totalPercentTime, 3));
+            TraceUtil.renderRight(out, 10, put1(_, "%d", count));
+            out.outln();
+        }
+
+        if (maxDepth == -1 || depth < maxDepth) {
+            for (child in childArray) {
+                if (child != null) child.render(out, m, depth + 1, totalPercentTime, maxDepth);
+            }
+        }
+    }
+    private def byTime(a: ProfileTreeNode, b: ProfileTreeNode) -> bool {
+        return b.totalTime < a.totalTime;
+    }
+
+    private def renderHeader(out: TraceBuilder) {
+        var l = out.length;
+        out.puts(TermColors.CYAN);
+        out.puts("*");
+        out.puts(TermColors.DEFAULT);
+        out.pad(' ', l + 50);
+        TraceUtil.renderRight(out, 12, puts(_, "nested"));
+        TraceUtil.renderRight(out, 12, puts(_, "self"));
+        TraceUtil.renderRight(out, 10, puts(_, "%"));
+        TraceUtil.renderRight(out, 10, puts(_, "count"));
+        out.outln();
+    }
+    
+    
+    private def puts(out: TraceBuilder, s: string) {
+        out.puts(s);
+    }
+    private def put1<T1>(out: TraceBuilder, fmt: string, p1: T1) {
+        out.put1(fmt, p1);
+    }
+}
+
+class ProfileTree {
+    var rootNode: ProfileTreeNode;
+    var curNode: ProfileTreeNode;
+
+    new() {
+        rootNode = ProfileTreeNode.new(null, null);
+        curNode = rootNode;
+    }
+    def enterFunc(func: WasmFunction) {
+        curNode = curNode.getOrCreateChild(func);
+        curNode.enter();
+    }
+    def exitFunc() {
+        curNode.exit();
+        curNode = curNode.parent;
+    }
+}
+
+class ProfileMonitor extends Monitor {
+    var tree: ProfileTree;
```

Comment:
The indentation is a little off in this file. I always use one tab per indentation level.